File: frontend/src/services/userService.js
/**
 * User service
 * Handles user management operations
 */

import api from './api';

const USER_ENDPOINTS = {
  BASE: '/users',
  BY_ID: (id) => `/users/${id}`,
  TOGGLE_STATUS: (id) => `/users/${id}/status`,
};

/**
 * Gets all users
 * @returns {Promise} Promise containing array of users
 */
export const getAllUsers = async () => {
  try {
    const response = await api.get(USER_ENDPOINTS.BASE);
    return response.data;
  } catch (error) {
    throw error.response?.data || { message: 'Failed to fetch users' };
  }
};

/**
 * Gets a user by ID
 * @param {number} id - User ID
 * @returns {Promise} Promise containing user data
 */
export const getUserById = async (id) => {
  try {
    const response = await api.get(USER_ENDPOINTS.BY_ID(id));
    return response.data;
  } catch (error) {
    throw error.response?.data || { message: 'Failed to fetch user' };
  }
};

/**
 * Creates a new user
 * @param {Object} userData - User data
 * @returns {Promise} Promise containing created user data
 */
export const createUser = async (userData) => {
  try {
    const response = await api.post(USER_ENDPOINTS.BASE, userData);
    return response.data;
  } catch (error) {
    throw error.response?.data || { message: 'Failed to create user' };
  }
};

/**
 * Updates an existing user
 * @param {number} id - User ID
 * @param {Object} userData - User data
 * @returns {Promise} Promise containing updated user data
 */
export const updateUser = async (id, userData) => {
  try {
    const response = await api.put(USER_ENDPOINTS.BY_ID(id), userData);
    return response.data;
  } catch (error) {
    throw error.response?.data || { message: 'Failed to update user' };
  }
};

/**
 * Deletes a user
 * @param {number} id - User ID
 * @returns {Promise} Promise containing delete response
 */
export const deleteUser = async (id) => {
  try {
    const response = await api.delete(USER_ENDPOINTS.BY_ID(id));
    return response.data;
  } catch (error) {
    throw error.response?.data || { message: 'Failed to delete user' };
  }
};

/**
 * Toggles user active status
 * @param {number} id - User ID
 * @param {boolean} active - New active status
 * @returns {Promise} Promise containing updated user data
 */
export const toggleUserStatus = async (id, active) => {
  try {
    const response = await api.patch(USER_ENDPOINTS.TOGGLE_STATUS(id), null, {
      params: { active },
    });
    return response.data;
  } catch (error) {
    throw error.response?.data || { message: 'Failed to toggle user status' };
  }
};

File: backend/src/main/java/com/adrs/service/UserService.java
package com.adrs.service;

import com.adrs.dto.AuthResponse;
import com.adrs.dto.LoginRequest;
import com.adrs.dto.UserRequest;
import com.adrs.dto.UserResponse;
import com.adrs.model.Province;
import com.adrs.model.User;

import java.util.List;

/**
 * Service interface for user-related operations.
 */
public interface UserService {

    /**
     * Authenticates a user and returns an authentication response.
     *
     * @param loginRequest the login request containing username and password
     * @return AuthResponse containing JWT token and user details
     */
    AuthResponse authenticateUser(LoginRequest loginRequest);

    /**
     * Creates a new user in the system.
     *
     * @param userRequest the user creation request
     * @return UserResponse containing the created user details
     */
    UserResponse createUser(UserRequest userRequest);

    /**
     * Updates an existing user.
     *
     * @param id          the user ID
     * @param userRequest the user update request
     * @return UserResponse containing the updated user details
     */
    UserResponse updateUser(Long id, UserRequest userRequest);

    /**
     * Retrieves a user by ID.
     *
     * @param id the user ID
     * @return UserResponse containing the user details
     */
    UserResponse getUserById(Long id);

    /**
     * Retrieves all users in the system.
     *
     * @return a list of all users
     */
    List<UserResponse> getAllUsers();

    /**
     * Deletes a user by ID.
     *
     * @param id the user ID
     */
    void deleteUser(Long id);

    /**
     * Activates or deactivates a user account.
     *
     * @param id     the user ID
     * @param active the active status
     * @return UserResponse containing the updated user details
     */
    UserResponse toggleUserStatus(Long id, Boolean active);

    /**
     * Retrieves users filtered by province and optionally by role.
     *
     * @param province the province to filter by
     * @param role     the optional role to filter by (null for all roles)
     * @return a list of users matching the filters
     */
    List<UserResponse> getUsersByProvinceAndRole(Province province, User.Role role);
}
